lib/storage/data/external/AwsClient.js

```diff
@@ -140,78 +236,139 @@ class AwsClient {
-        stream._destroy = (err, cb) => {
-            log.debug('aborting GET request in progress', { objectGetInfo });
-            request.abort();
-            if (streamDestroy) {
-                streamDestroy(err, cb);
-            }
+            Range: range ? `bytes=${range[0]}-${range[1]}` : undefined,
         };
-        stream.on('error', err => {
-            let logLevel;
-            if (err.code === 'NotFound') {
-                logLevel = 'info';
-            } else {
-                logLevel = 'error';
-            }
-            logHelper(log, logLevel,
-                `error streaming data from ${this.type}`,
-                err, this._dataStoreName, this.clientType);
-        });
-        // Always call the callback asynchronously: the caller may
-        // destroy the stream with destroy(), which MUST be
-        // asynchronous wrt. request.createReadStream() to avoid
-        // socket leaks, notably because of the request.abort() call.
-        return process.nextTick(() => callback(null, stream));
+        const command = new GetObjectCommand(params);
+        const abortController = new AbortController();
+        this._client.send(command, { abortSignal: abortController.signal })
+            .then(data => {
+                const rawBody = data.Body;
+                const stream = this._normalizeSdkStream(rawBody);
+                if (!stream || typeof stream.on !== 'function') {
+                    const bodyType = rawBody?.constructor?.name || typeof rawBody;
+                    log.error('unsupported body type from AWS SDK getObject response', {
+                        method: 'AwsClient.get',
+                        backendType: this.clientType,
+                        dataStoreName: this._dataStoreName,
+                        bodyType,
+                    });
+                    return cbOnce(errorInstances.InternalError
+                        .customizeDescription('Unsupported response body type returned from AWS SDK'));
+                }
+
+                if (data.$metadata?.httpHeaders) {
+                    log.trace(`${this.type} GET request response headers`, {
+                        responseHeaders: data.$metadata.httpHeaders,
+                        backendType: this.clientType,
+                    });
+                }
+
+                let finished = false;
+                const originalDestroy = stream.destroy.bind(stream);
+
+                const abortRequest = err => {
+                    if (finished) {
+                        return stream;
+                    }
+                    finished = true;
+                    log.debug('aborting GET request in progress', { objectGetInfo });
+                    abortController.abort(err);
+                    if (rawBody && rawBody !== stream) {
+                        rawBody.destroy?.(err);
+                    }
+                    try {
+                        originalDestroy(err);
```

Comment:
should destroy be called on `abort()`? should it not instead be called only on `destroy()` ?
(not sure what the semantics of JS streams actually are)